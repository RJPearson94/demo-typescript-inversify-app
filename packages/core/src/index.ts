import '@src/constant/types';
import '@src/service';
import '@src/inversify.config';
