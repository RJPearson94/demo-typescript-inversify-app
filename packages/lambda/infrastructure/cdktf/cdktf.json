{
  "language": "typescript",
  "app": "ts-node bin/app.ts",
  "terraformProviders": [],
  "terraformModules": [],
  "context": {
    "excludeStackIdFromLogicalIds": "true",
    "allowSepCharsInLogicalIds": "true"
  }
}
