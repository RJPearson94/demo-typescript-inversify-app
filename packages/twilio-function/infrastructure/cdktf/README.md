# @rjpearson94/twilio-function-cdktf-infrastructure

This package is responsible for deploying the function to Twilio using [CDK for Terraform](https://github.com/hashicorp/terraform-cdk)

## Getting Started

To be able to test and run the function on real Twilio infrastructure you need the following:

- [Twilio Account](https://www.twilio.com/)
- [Yarn 2/ Berry](https://yarnpkg.com/)

## Commands

- synth (Generate Terraform JSON template)
- get (Download and generate the necessary constructs)
- test (Run all tests)

For the full commands list please see the [package.json](./package.json)

The commands can be run using the following

Open a command prompt in this directory and run

```sh
yarn <script>
```

Before the Twilio function resources can be managed via Terraform (creation, updating, destruction), the Twilio function javascript file must be built and stored in the dist folder of the twilio-function package. If the javascript file has not been built you will need to follow the instructions in the Twilio function [README](../../README.md)

Additionally, the Twilio provider constructs need to be autogenerated before the resources can be managed or the tests can run. To generate the CDKTF constructs please run the following command

```sh
yarn get
```

## Deploying to Twilio

To see what resources will be provisioned, change into the deployment folder and run the following:

```bash
yarn synth
```

This will show you all the resources that will be provisioned, including:

- [Functions](https://www.twilio.com/docs/runtime/functions)

If you want to provision these resources, please run the following command (Please note cost/ charges may be incurred):

```bash
yarn deploy
```

When you see the following:

```txt
Do you want to perform these actions?
  CDK for Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value:
```

type `yes` and hit the return/ enter key. Alternatively, if you don't want the manual accept process you can run the following command:

```bash
yarn deploy --auto-approve
```

The resources should be provisioned

## Remove Resources from Twilio

Once you are finished with the resources you can remove all the provisioned resources by running the following command (from the deployment folder):

```bash
yarn destroy
```

When you see the following

```txt
Do you want to perform these actions?
 CDK for Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value:
```

type `yes` and hit the return/ enter key. Alternatively, if you don't want the manual accept process you can run the following command:

```bash
yarn destroy --auto-approve
```

Then all the resources (which CDK for Terraform has state information for) should be removed.
