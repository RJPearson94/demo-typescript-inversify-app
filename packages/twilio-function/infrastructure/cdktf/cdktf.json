{
  "language": "typescript",
  "app": "ts-node bin/app.ts",
  "terraformProviders": ["RJPearson94/twilio@>= 0.8.0"],
  "terraformModules": [],
  "context": {
    "excludeStackIdFromLogicalIds": "true",
    "allowSepCharsInLogicalIds": "true"
  }
}
