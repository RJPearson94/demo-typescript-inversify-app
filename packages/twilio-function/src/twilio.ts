import 'reflect-metadata';
import 'source-map-support/register';

import { TYPES } from './constant/types';
import { GreetingController } from './controller';
import { container } from './inversify.config';
import { MessageResponse } from './lib';

const twilioHandler = async (_context: any, _event: any, callback: (err: Error, resp: MessageResponse) => void): Promise<void> => {
  const greetingController = container.get<GreetingController>(TYPES.GreetingController);
  const helloResponse = greetingController.greet();

  callback(null, {
    message: helloResponse
  });
};

export const handler = twilioHandler;
