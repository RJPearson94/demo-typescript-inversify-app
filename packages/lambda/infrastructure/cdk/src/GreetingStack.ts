import { CfnOutput, Construct, Duration, RemovalPolicy, Stack, StackProps } from '@aws-cdk/core';
import { Function, Runtime, Code, Tracing, Alias, IFunction, IAlias } from '@aws-cdk/aws-lambda';
import { LogGroup, RetentionDays } from '@aws-cdk/aws-logs';
import { IRestApi, SpecRestApi, ApiDefinition, RateLimitedApiKey, Period, IApiKey } from '@aws-cdk/aws-apigateway';
import { ServicePrincipal } from '@aws-cdk/aws-iam';
import openApiTemplate from './templates/openapi.tpl.json';

export interface FunctionResources {
  func: IFunction;
  alias: IAlias;
}

export interface ApiGatewayResources {
  restApi: IRestApi;
  apiKey: IApiKey;
}

export type GreetingStackProps = {
  resourcePrefix?: string;
  api?: {
    stageName?: string;
    quota?: {
      limit?: number;
      period?: Period;
    };
  };
  lambda?: {
    alias?: string;
  };
} & StackProps;

export class GreetingStack extends Stack {
  public readonly apiGatewayURL: CfnOutput;
  public readonly apiKeyArn: CfnOutput;

  private resourcePrefix: string;

  public constructor(scope: Construct, private id: string, private props?: GreetingStackProps) {
    super(scope, id, props);

    this.resourcePrefix = props?.resourcePrefix ? `${props?.resourcePrefix}-` : '';

    const { alias } = this.createFunction();
    const { restApi, apiKey } = this.createAPIGateway(alias);

    alias.addPermission('AllowApiGatewayInvoke', {
      principal: new ServicePrincipal('apigateway.amazonaws.com'),
      action: 'lambda:InvokeFunction',
      sourceArn: restApi.deploymentStage.restApi.arnForExecuteApi()
    });

    this.apiGatewayURL = new CfnOutput(this, `APIGatewayURL`, {
      value: restApi.deploymentStage.urlForPath()
    });
    this.apiKeyArn = new CfnOutput(this, `APIKeyARN`, {
      value: apiKey.keyArn
    });
  }

  private createFunction(): FunctionResources {
    const functionName = `${this.resourcePrefix}${this.id}`;

    const logGroup = new LogGroup(this, `FunctionLogGroup`, {
      logGroupName: `/aws/lambda/${functionName}`,
      retention: RetentionDays.ONE_WEEK,
      removalPolicy: RemovalPolicy.DESTROY
    });

    const func = new Function(this, `Function`, {
      code: Code.fromAsset('../../dist/lambda.zip'),
      functionName,
      handler: 'main.handler',
      runtime: Runtime.NODEJS_14_X,
      tracing: Tracing.ACTIVE,
      timeout: Duration.seconds(5)
    });

    logGroup.grantWrite(func);

    const alias = new Alias(this, `FunctionAlias`, {
      aliasName: this.props?.lambda?.alias || 'stable',
      version: func.currentVersion
    });

    return {
      func,
      alias
    };
  }

  private createAPIGateway(alias: IAlias): ApiGatewayResources {
    const template = JSON.stringify(openApiTemplate).replace('${lambda_arn}', alias.functionArn);

    const apiGatewayName = `${this.resourcePrefix}${this.id}`;

    const restApi = new SpecRestApi(this, `ApiGateway`, {
      apiDefinition: ApiDefinition.fromInline(JSON.parse(template)),
      restApiName: `${apiGatewayName}-rest-api`,
      deploy: true,
      cloudWatchRole: false,
      deployOptions: {
        tracingEnabled: true,
        stageName: this.props?.api?.stageName || 'test'
      }
    });

    const apiKey = new RateLimitedApiKey(this, `ApiKey`, {
      apiKeyName: `${apiGatewayName}-api-key`,
      resources: [restApi],
      apiStages: [
        {
          api: restApi,
          stage: restApi.deploymentStage
        }
      ],
      quota: {
        limit: 100,
        period: Period.MONTH,
        ...(this.props?.api?.quota || {})
      }
    });

    return {
      restApi,
      apiKey
    };
  }
}
