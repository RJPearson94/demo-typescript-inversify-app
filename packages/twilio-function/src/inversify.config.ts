import { container } from '@rjpearson94/core';

import { TYPES } from './constant/types';
import { GreetingController, GreetingControllerImpl } from './controller';

container.bind<GreetingController>(TYPES.GreetingController).to(GreetingControllerImpl);

export { container };
