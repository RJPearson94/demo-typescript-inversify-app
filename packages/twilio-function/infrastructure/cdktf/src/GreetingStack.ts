import { Construct } from 'constructs';
import { TerraformStack, TerraformOutput } from 'cdktf';
import { ServerlessBuild, ServerlessDeployment, ServerlessEnvironment, ServerlessFunction, ServerlessService } from '../.gen/providers/twilio';
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';

export type GreetingStackProps = {
  resourceSuffix?: string;
};

export class GreetingStack extends TerraformStack {
  public readonly functionURL: TerraformOutput;

  private resourceSuffix: string;

  constructor(scope: Construct, private name: string, props?: GreetingStackProps) {
    super(scope, name);

    this.resourceSuffix = props?.resourceSuffix ? `-${props?.resourceSuffix}` : '';

    const service = this.createService();
    const func = this.createFunction(service);
    const build = this.createBuild(service, func);
    const environment = this.createEnvironment(service);
    this.createDeployment(service, environment, build);

    this.functionURL = new TerraformOutput(this, `function_url`, {
      value: `https://${environment.domainName}${func.path}`
    });
  }

  private createService(): ServerlessService {
    return new ServerlessService(this, 'service', {
      uniqueName: `${this.name}${this.resourceSuffix}`,
      friendlyName: `${this.name}${this.resourceSuffix}`,
      includeCredentials: false,
      uiEditable: false
    });
  }

  private createFunction(service: ServerlessService): ServerlessFunction {
    const source = '../../../dist/index.js';
    const fileContents = fs.readFileSync(path.resolve(__dirname, source));

    return new ServerlessFunction(this, 'function', {
      serviceSid: service.sid,
      friendlyName: `${this.name}${this.resourceSuffix}`,

      source,
      sourceHash: crypto.createHash('sha256').update(fileContents).digest('base64'),

      contentType: 'application/javascript',
      path: '/inversify',
      visibility: 'public'
    });
  }

  private createBuild(service: ServerlessService, func: ServerlessFunction): ServerlessBuild {
    return new ServerlessBuild(this, 'build', {
      serviceSid: service.sid,

      functionVersion: [
        {
          sid: func.latestVersionSid
        }
      ],

      runtime: 'node12',

      dependencies: {
        twilio: '3.57.0',
        fs: '0.0.1-security',
        lodash: '4.17.21',
        util: '0.12.3',
        xmldom: '0.4.0',
        '@twilio/runtime-handler': '1.0.1'
      },

      polling: [
        {
          enabled: true
        }
      ],

      lifecycle: {
        createBeforeDestroy: true
      }
    });
  }

  private createEnvironment(service: ServerlessService): ServerlessEnvironment {
    return new ServerlessEnvironment(this, 'environment', {
      serviceSid: service.sid,
      uniqueName: `dev${this.resourceSuffix}`
    });
  }

  private createDeployment(service: ServerlessService, environment: ServerlessEnvironment, build: ServerlessBuild): ServerlessDeployment {
    return new ServerlessDeployment(this, 'deployment', {
      serviceSid: service.sid,
      environmentSid: environment.sid,
      buildSid: build.sid,

      lifecycle: {
        createBeforeDestroy: true
      }
    });
  }
}
