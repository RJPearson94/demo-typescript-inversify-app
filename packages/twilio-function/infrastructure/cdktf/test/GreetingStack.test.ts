import { Testing } from 'cdktf';
import { GreetingStack } from '../src/GreetingStack';

test('create greeting stack', () => {
  const app = Testing.app();
  const stack = new GreetingStack(app, 'greeting-stack');
  expect(Testing.synth(stack)).toMatchSnapshot();
});

test('create greeting stack with suffix', () => {
  const app = Testing.app();
  const stack = new GreetingStack(app, 'greeting-stack', {
    resourceSuffix: 'test'
  });
  expect(Testing.synth(stack)).toMatchSnapshot();
});
