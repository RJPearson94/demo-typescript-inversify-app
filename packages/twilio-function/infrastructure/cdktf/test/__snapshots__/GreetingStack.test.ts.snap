// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create greeting stack 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"greeting-stack\\"
    }
  },
  \\"resource\\": {
    \\"twilio_serverless_service\\": {
      \\"service\\": {
        \\"friendly_name\\": \\"greeting-stack\\",
        \\"include_credentials\\": false,
        \\"ui_editable\\": false,
        \\"unique_name\\": \\"greeting-stack\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/service\\",
            \\"uniqueId\\": \\"service\\"
          }
        }
      }
    },
    \\"twilio_serverless_function\\": {
      \\"function\\": {
        \\"content_type\\": \\"application/javascript\\",
        \\"friendly_name\\": \\"greeting-stack\\",
        \\"path\\": \\"/inversify\\",
        \\"service_sid\\": \\"\${twilio_serverless_service.service.sid}\\",
        \\"source\\": \\"../../../dist/index.js\\",
        \\"source_hash\\": \\"filebase64sha256(../../../dist/index.js)\\",
        \\"visibility\\": \\"public\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/function\\",
            \\"uniqueId\\": \\"function\\"
          }
        }
      }
    },
    \\"twilio_serverless_build\\": {
      \\"build\\": {
        \\"dependencies\\": {
          \\"twilio\\": \\"3.57.0\\",
          \\"fs\\": \\"0.0.1-security\\",
          \\"lodash\\": \\"4.17.21\\",
          \\"util\\": \\"0.12.3\\",
          \\"xmldom\\": \\"0.4.0\\",
          \\"@twilio/runtime-handler\\": \\"1.0.1\\"
        },
        \\"runtime\\": \\"node12\\",
        \\"service_sid\\": \\"\${twilio_serverless_service.service.sid}\\",
        \\"function_version\\": [
          {
            \\"sid\\": \\"\${twilio_serverless_function.function.latest_version_sid}\\"
          }
        ],
        \\"polling\\": [
          {
            \\"enabled\\": true
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/build\\",
            \\"uniqueId\\": \\"build\\"
          }
        }
      }
    },
    \\"twilio_serverless_environment\\": {
      \\"environment\\": {
        \\"service_sid\\": \\"\${twilio_serverless_service.service.sid}\\",
        \\"unique_name\\": \\"dev\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/environment\\",
            \\"uniqueId\\": \\"environment\\"
          }
        }
      }
    },
    \\"twilio_serverless_deployment\\": {
      \\"deployment\\": {
        \\"build_sid\\": \\"\${twilio_serverless_build.build.sid}\\",
        \\"environment_sid\\": \\"\${twilio_serverless_environment.environment.sid}\\",
        \\"service_sid\\": \\"\${twilio_serverless_service.service.sid}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/deployment\\",
            \\"uniqueId\\": \\"deployment\\"
          }
        }
      }
    }
  },
  \\"output\\": {
    \\"function_url\\": {
      \\"value\\": \\"https://\${twilio_serverless_environment.environment.domain_name}\${twilio_serverless_function.function.path}\\"
    }
  }
}"
`;

exports[`create greeting stack with suffix 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"greeting-stack\\"
    }
  },
  \\"resource\\": {
    \\"twilio_serverless_service\\": {
      \\"service\\": {
        \\"friendly_name\\": \\"greeting-stack-test\\",
        \\"include_credentials\\": false,
        \\"ui_editable\\": false,
        \\"unique_name\\": \\"greeting-stack-test\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/service\\",
            \\"uniqueId\\": \\"service\\"
          }
        }
      }
    },
    \\"twilio_serverless_function\\": {
      \\"function\\": {
        \\"content_type\\": \\"application/javascript\\",
        \\"friendly_name\\": \\"greeting-stack-test\\",
        \\"path\\": \\"/inversify\\",
        \\"service_sid\\": \\"\${twilio_serverless_service.service.sid}\\",
        \\"source\\": \\"../../../dist/index.js\\",
        \\"source_hash\\": \\"filebase64sha256(../../../dist/index.js)\\",
        \\"visibility\\": \\"public\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/function\\",
            \\"uniqueId\\": \\"function\\"
          }
        }
      }
    },
    \\"twilio_serverless_build\\": {
      \\"build\\": {
        \\"dependencies\\": {
          \\"twilio\\": \\"3.57.0\\",
          \\"fs\\": \\"0.0.1-security\\",
          \\"lodash\\": \\"4.17.21\\",
          \\"util\\": \\"0.12.3\\",
          \\"xmldom\\": \\"0.4.0\\",
          \\"@twilio/runtime-handler\\": \\"1.0.1\\"
        },
        \\"runtime\\": \\"node12\\",
        \\"service_sid\\": \\"\${twilio_serverless_service.service.sid}\\",
        \\"function_version\\": [
          {
            \\"sid\\": \\"\${twilio_serverless_function.function.latest_version_sid}\\"
          }
        ],
        \\"polling\\": [
          {
            \\"enabled\\": true
          }
        ],
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/build\\",
            \\"uniqueId\\": \\"build\\"
          }
        }
      }
    },
    \\"twilio_serverless_environment\\": {
      \\"environment\\": {
        \\"service_sid\\": \\"\${twilio_serverless_service.service.sid}\\",
        \\"unique_name\\": \\"dev-test\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/environment\\",
            \\"uniqueId\\": \\"environment\\"
          }
        }
      }
    },
    \\"twilio_serverless_deployment\\": {
      \\"deployment\\": {
        \\"build_sid\\": \\"\${twilio_serverless_build.build.sid}\\",
        \\"environment_sid\\": \\"\${twilio_serverless_environment.environment.sid}\\",
        \\"service_sid\\": \\"\${twilio_serverless_service.service.sid}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"greeting-stack/deployment\\",
            \\"uniqueId\\": \\"deployment\\"
          }
        }
      }
    }
  },
  \\"output\\": {
    \\"function_url\\": {
      \\"value\\": \\"https://\${twilio_serverless_environment.environment.domain_name}\${twilio_serverless_function.function.path}\\"
    }
  }
}"
`;
